src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,485 @@
+import boto3
+import josepy as jose
+import json
+import requests
+import time
+
+from middlewared.schema import Bool, Dict, Int, Str, ValidationErrors
+from middlewared.service import accepts, CallError, CRUDService, filterable, private
+from middlewared.validators import validate_attributes
+
+from acme import client, messages
+from botocore import exceptions as boto_exceptions
+from botocore.errorfactory import BaseClientExceptions as boto_BaseClientException
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+
+# TODO: See what can be done to respect rate limits
+
+
+class ACMERegistrationService(CRUDService):
+
+    class Config:
+        datastore = 'system.acmeregistration'
+        datastore_extend = 'acme.registration.register_extend'
+        namespace = 'acme.registration'
+        private = True
+
+    @filterable
+    async def query(self, filters=None, options=None):
+
+        """
+
+        Query Registered ACME Servers for user "root"
+
+        """
+        for filter in filters or []:
+            if len(filter) == 3 and filter[0] == 'directory':
+                # Let's normalize directory uri for query
+                filter[2] += '/' if filter[2][-1] != '/' else ''
+
```

Comment:
This feels strange. Why are we normalizing user input?

Reply:
We do this because in database, we store each directory uri with a trailing slash. This is to make sure that when a filter for uri is applied, a trailing slash is present

Reply:
If the uri is with a trailing slash why are we allowing to filter without it? Seems like a lot of work for something that is already behaving as it should?